infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/logreader/StreamsSnapshotReader.java

```diff
@@ -56,68 +67,103 @@
-                (MessageType.SNAPSHOT_MESSAGE, topologyConfigId, snapshotRequestId, currentMsgTs,
-                preMsgTs, snapshotTimestamp, sequence, buf.array());
         preMsgTs = currentMsgTs;
         sequence++;
-        log.debug("Generate TxMsg {}", txMsg.getMetadata());
+
+        log.trace("txMsg {} deepsize sizeInBytes {} entryList.sizeInByres {}  with numEntries {} deepSize sizeInBytes {}",
+                txMsg.getMetadata(), MetricsUtils.sizeOf.deepSizeOf(txMsg), entryList.getSizeInBytes(), entryList.getSmrEntries().size(), MetricsUtils.sizeOf.deepSizeOf(entryList.smrEntries));
+
         return txMsg;
     }
 
     /**
-     * Read numEntries from the current stream.
+     * Read log data from the current stream until the sum of all SMR entries's sizeInBytes reaches the maxDataSizePerMsg.
      * @param stream
-     * @param numEntries
      * @return
      */
-    private List<SMREntry> next(OpaqueStreamIterator stream, int numEntries) {
-        //if it is the end of the stream, set an end of stream mark, the current
-        List<SMREntry> list = new ArrayList<>();
+    private SMREntryList next(OpaqueStreamIterator stream) {
+        List<SMREntry> smrList = new ArrayList<>();
+        int currentMsgSize = 0;
+
         try {
-            while (stream.iterator.hasNext() && list.size() < numEntries) {
-                OpaqueEntry entry = (OpaqueEntry) stream.iterator.next();
-                stream.maxVersion = Math.max(stream.maxVersion, entry.getVersion());
-                list.addAll(entry.getEntries().get(stream.uuid));
+            while (currentMsgSize < maxDataSizePerMsg) {
+                if (lastEntry != null) {
+                    List<SMREntry> smrEntries = lastEntry.getEntries().get(stream.uuid);
+                    if (smrEntries != null) {
+                        int currentEntrySize = ReaderUtility.calculateSize(smrEntries);
+
+                        if (currentEntrySize > MAX_LOG_REPLICATION_DATA_MSG_SIZE_SUPPORTED) {
+                            log.error("The current entry size {} is bigger than the maxDataSizePerMsg {} supported", currentEntrySize, MAX_LOG_REPLICATION_DATA_MSG_SIZE_SUPPORTED);
+                            throw new IllegalSnapshotEntrySizeException(" The snapshot entry is bigger than the system supported");
+                        } else if (currentEntrySize > maxDataSizePerMsg) {
+                            observeBiggerMsg.setValue(observeBiggerMsg.getValue()+1);
+                            log.warn("The current entry size {} is bigger than the configured maxDataSizePerMsg {}",
+                                    currentEntrySize, maxDataSizePerMsg);
+                        }
+
+                        // Skip append this entry in this message. Will process it first at the next round.
+                        if (currentEntrySize + currentMsgSize > maxDataSizePerMsg && currentMsgSize != 0) {
```

Comment:
why do we need currentMsgSize != 0? If we already checked that currentEntrySize is bigger previously and we throw an exception.

Reply:
If it is the only entry and its size is bigger than the configured but smaller than the max_supported, we still continue the replication , right? I thought we have discussed this before.